### PR TITLE
initial esp support

### PR DESCRIPTION
**Pull request by kaspar030**
This PR adds initial esp32 support.

Currently, this adds support for a random esp32c3 (RISC-V based) board, the necessary infra for esp-hal, and esp-wifi.

There's no preemptive scheduling (riot-rs-threads) [yet](https://github.com/future-proof-iot/RIOT-rs/issues/157).
A single thread-mode executor is used.

*edit*

I'm testing with:

    CONFIG_WIFI_NETWORK="your-ssid" CONFIG_WIFI_PASSWORD="your-pwd" laze build -Cexamples/embassy-http-server -b ai-c3 -s debug-console -d silent-panic run
    
 This implies connecting to wifi but using the hard-coded IP address. to make the esp use dhcp, remove the "override-network-config" from `examples/embassy-http-server/Cargo.toml`.
 
 Ah, it needs some toolchain support.
 At least `laze b install-toolchain` and `cargo install espflash`.

### CODE REVIEW

File: src/riot-rs-embassy/src/lib.rs [hunk 1]
```diff
@@ -56,9 +57,18 @@ pub static EMBASSY_TASKS: [Task] = [..];
...
+    let p = crate::arch::OptionalPeripherals::from(p);
+
+    #[cfg(any(context = "nrf52", context = "rp2040"))]
     EXECUTOR.start(arch::SWI);
-    EXECUTOR.spawner().spawn(init_task(p)).unwrap();
+
+    #[cfg(context = "esp")]
+    EXECUTOR.start(arch::interrupt::Priority::Priority1);
+
+    EXECUTOR.spawner().must_spawn(init_task(p));
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs:71]: hm, `must_spawn()` is what upstream uses. I think the main benefit is that embassy's `unwrap!()` uses the [`defmt` version](https://docs.rs/defmt/latest/defmt/macro.unwrap.html) if that is available.

File: src/riot-rs-embassy/src/lib.rs [hunk 2]
```diff
@@ -56,9 +57,18 @@ pub static EMBASSY_TASKS: [Task] = [..];
 #[distributed_slice(riot_rs_rt::INIT_FUNCS)]
 pub(crate) fn init() {
     riot_rs_rt::debug::println!("riot-rs-embassy::init()");
-    let p = arch::OptionalPeripherals::from(arch::init(Default::default()));
+    let p = arch::init(Default::default());
+
+    #[cfg(not(context = "esp"))]
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs:62]: Yeah, makes sense. (The way to this version had way more special caseing already. :smile: )

File: src/riot-rs-embassy/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs]: ... let's discuss this outside of this PR ...

[the remark this answers is not in the document]
Comment (bergzand) [reply src/riot-rs-embassy/src/lib.rs]: Could the result of the offline discussion be added to the documentation if it is also applicable elsewhere in the code base?

File: laze-project.yml [hunk 3]
```diff
@@ -409,6 +431,18 @@ modules:
       - rpi-pico
       - rpi-pico-w
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply laze-project.yml:432]: it has, but, we'd only list it here if it is actually supported. (which it isn't yet)

File: src/riot-rs-rt/src/esp.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#[esp_hal::entry]
+fn main() -> ! {
+    super::startup();
+}
+
+pub fn init() {
+    //    esp_println::logger::init_logger(log::LevelFilter::Debug);
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-rt/src/esp.rs:7]: ah. I think we'll need to provide unified logging configuration. But I used these while debugging.
This one's removed now.

File: src/riot-rs-embassy/src/arch/esp.rs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    let timer_group0 = TimerGroup::new(peripherals.TIMG0.take().unwrap(), &clocks);
+    embassy::init(&clocks, timer_group0);
+
+    peripherals
+}
+
+// #[cfg(feature = "usb")]
+// use embassy_rp::{bind_interrupts, peripherals::USB, usb::InterruptHandler};
+
+// // rp2040 usb start
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/arch/esp.rs:45]: not really, I can copy&paste it again.

File: src/riot-rs-embassy/src/lib.rs [hunk 6]
```diff
@@ -20,7 +21,7 @@ pub mod usb;
 #[cfg(feature = "net")]
 pub mod network;
 
-#[cfg(feature = "wifi-cyw43")]
+#[cfg(any(feature = "wifi-cyw43", feature = "esp-wifi"))]
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs]: > major: I think we should instead introduce a `wifi` feature and make those two chip-specific features enable it.

agreed & done.

[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs]: > Is the initialization process different when behaving as a Wi-Fi client and as an AP? Should we have two different features for that?

Yes it is. Let's tackle that for all (both) implementations in another PR? Currently, we only support client mode for both.

File: src/riot-rs-embassy/src/lib.rs [hunk 7]
```diff
@@ -38,6 +39,9 @@ use usb::ethernet::NetworkDevice;
 #[cfg(feature = "wifi-cyw43")]
 use wifi::cyw43::NetworkDevice;
 
+#[cfg(feature = "esp-wifi")]
+use wifi::esp_wifi::NetworkDevice;
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs:43]: done

File: src/riot-rs-embassy/src/wifi/esp_wifi.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            controller.set_configuration(&client_config).unwrap();
+            println!("Starting wifi");
+            controller.start().await.unwrap();
+            println!("Wifi started!");
+        }
+        println!("About to connect...");
+
+        match controller.connect().await {
+            Ok(_) => println!("Wifi connected!"),
+            Err(e) => {
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/wifi/esp_wifi.rs:55]: that I'll leave open for discussion. :)

File: src/riot-rs-embassy/src/wifi/esp_wifi.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
+use esp_wifi::{wifi::WifiStaDevice, EspWifiInitialization};
+use once_cell::sync::OnceCell;
+
+use crate::{arch::OptionalPeripherals, Spawner};
+
+pub use esp_wifi::wifi::{WifiController, WifiDevice};
+
+pub type NetworkDevice = WifiDevice<'static, WifiStaDevice>;
+
+pub static WIFI_INIT: OnceCell<EspWifiInitialization> = OnceCell::new();
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/wifi/esp_wifi.rs:10]: added a comment

File: src/riot-rs-rt/src/debug.rs [hunk 10]
```diff
@@ -31,6 +31,19 @@ mod backend {
...
+#[cfg(all(feature = "debug-console", context = "esp"))]
+mod backend {
+    pub use esp_println::{print, println};
+    pub const EXIT_SUCCESS: Result<(), ()> = Ok(());
+    pub const EXIT_FAILURE: Result<(), ()> = Err(());
+    pub fn exit(_code: Result<(), ()>) {
+        loop {}
+    }
+    pub fn init() {
+        //        esp_println::logger::init_logger_from_env();
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-rt/src/debug.rs]: It does. hmm, we'll need to unify logging ... I'll re-add this for now.

File: src/riot-rs-rt/src/cortexm.rs [hunk 11]
```diff
@@ -26,7 +26,7 @@ pub fn ipsr_isr_number_to_str(isr_number: usize) -> &'static str {
 /// Extra verbose Cortex-M HardFault handler
 ///
 /// (copied from Tock OS)
-#[allow(non_snake_case)]
+#[allow(non_snake_case, unsafe_op_in_unsafe_fn)]
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-rt/src/cortexm.rs:29]: https://github.com/future-proof-iot/RIOT-rs/pull/162

**Comment by kaspar030** [comment]
@ROMemories I think I got most of your review comments. (sorry for already squashing, I had trouble absorbing the fixes into the right commits)

**Comment by elenaf9** [comment]
How is exception/ interrupt handling implementing in `esp-hal`?

I have only started looking into this, but it might be that this PR will conflict with a `riscv` implementation in `riot-rs-threading`.

Interrupts/ traps in RISC-V are handled by writing the address of the trap entry function into a CSR (`mtvec`). 
From what I understand/ read so far, I think `esp-hal` and `riot-rs-rt` (used in `riot-rs-threading`) both setup the `mtvec` register with their own trap handling logic [^1] [^2], which could cause problems when both are enabled.
Have you looked into this? I believe `esp-hal` offers some API for implementing a custom interrupt handler [^3], which might help resolving this conflict, but I don't know how that would impact the rest of its functionality.

[^1]: https://github.com/esp-rs/esp-hal/blob/85ee3807085ceff0e4e03b5ad053baebd7c8c109/esp-riscv-rt/src/lib.rs#L251C5-L251C11
[^2]: https://github.com/rust-embedded/riscv/blob/555441cbcd4c4d3dcee9185de56694a7cedec163/riscv-rt/src/asm.rs#L76-L81
[^3]: https://github.com/esp-rs/esp-hal/blob/85ee3807085ceff0e4e03b5ad053baebd7c8c109/esp-hal/ld/esp32c3/esp32c3.x#L20-L22

File: laze-project.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaspar030) [reply laze-project.yml:463]: (as in, I'll try to wire up the device USB port as a follow up PR)

File: src/riot-rs-embassy/src/wifi/esp_wifi.rs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+#[embassy_executor::task]
+async fn connection(mut controller: WifiController<'static>) {
+    use riot_rs_rt::debug::println;
+
+    use embassy_time::{Duration, Timer};
+    use esp_wifi::wifi::{ClientConfiguration, Configuration, WifiEvent, WifiState};
+
+    println!("start connection task");
+    println!("Device capabilities: {:?}", controller.get_capabilities());
+    loop {
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/wifi/esp_wifi.rs:36]: yeah ... the wifi controlling needs to be unified ... Folo! :)

File: src/riot-rs-embassy/src/lib.rs [hunk 13]
```diff
@@ -48,19 +52,36 @@ pub mod sendcell;
 
 pub type Task = fn(&Spawner, &mut arch::OptionalPeripherals);
 
-pub static EXECUTOR: arch::Executor = arch::Executor::new();
-
 #[distributed_slice]
 pub static EMBASSY_TASKS: [Task] = [..];
 
+#[cfg(feature = "executor-interrupt")]
+pub static EXECUTOR: arch::Executor = arch::Executor::new();
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-embassy/src/lib.rs:55]: EXECUTOR is used in `examples/embassy`. that's kinda broken now, but let's postpone fixing (and un-pubbing EXECUTOR)

**Pull request merged**